bokehjs/src/lib/core/visuals.ts

```diff
@@ -185,43 +184,37 @@ export abstract class ContextProperties {
   /** @prototype */
   attrs: string[]
 
-  readonly cache: {[key: string]: any} = {}
+  protected readonly cache: {[key: string]: any} = {}
 
-  abstract get doit(): boolean
+  private readonly _props: p.Property<unknown>[]
 
-  constructor(readonly obj: HasProps, readonly prefix: string = "") {
-    for (const attr of this.attrs)
-      (this as any)[attr] = obj.properties[prefix + attr]
+  *[Symbol.iterator](): Generator<p.Property<unknown>, void, undefined> {
+    yield* this._props
   }
 
-  warm_cache(source?: ColumnarDataSource, all_indices?: Indices): void {
+  constructor(readonly obj: View, readonly prefix: string = "") {
+    const self = this as any
+    this._props = []
     for (const attr of this.attrs) {
-      const prop = this.obj.properties[this.prefix + attr]
-      if (prop.spec.value !== undefined) // TODO (bev) better test?
-        this.cache[attr] = prop.spec.value
-      else if (source != null && prop instanceof p.VectorSpec) {
-        const array = prop.array(source)
-        const subarray = all_indices != null ? all_indices.select(array) : array
-        this.cache[attr + "_array"] = subarray
-      } else
-        throw new Error("source is required with a vectorized visual property")
+      const prop = obj.model.properties[prefix + attr]
+      self[attr] = prop
+      this._props.push(prop)
     }
   }
 
-  cache_select(attr: string, i: number): any {
-    const prop = this.obj.properties[this.prefix + attr]
-    let value: any
+  cache_select(prop: p.Property<unknown>, i: number): any {
     if (prop.spec.value !== undefined) // TODO (bev) better test?
```

Comment:
didn't you add a predicate for fields/values recently?

Reply:
I did. I will change this.